File: cloud/src/main/java/server/model/crdts/AWORMap.java
package server.model.crdts;

import org.json.JSONArray;
import org.json.JSONObject;
import server.model.utils.Pair;

import server.model.Product;
import server.model.crdts.DotContext;

import java.util.*;

public class AWORMap {
    private String id; // This is most likely deprecated
    /**
     * Map that stores [element_id, dot_value] => element
     */
    private HashMap<Pair<String, Integer>, Product> map;
    /**
     * DotContext that stores element_id => dot_value
     */
    private DotContext cc;

    public AWORMap() {
        map = new HashMap<>();
        cc = new DotContext();
    }

    /**
     * Get the map
     *
     * @return Map
     */
    public Map<Pair<String, Integer>, Product> getMap() {
        return map;
    }

    /**
     * Get the causal context
     *
     * @return Causal context
     */
    public DotContext getCC() {
        return cc;
    }

    /**
     * Get the entries of the map
     *
     * @return Set of map entries
     */
    public Set<Map.Entry<Pair<String, Integer>, Product>> entries() {
        return map.entrySet();
    }

    public Set<Pair<String, Integer>> elements() {
        return map.keySet();
    }

    /**
     * Get the values of the map that are in the causal context
     *
     * @return Iterator of values
     */
    public Collection<Product> values() {
        Map<Pair<String, Integer>, Product> filteredMap = new HashMap<>();
        Map<String, Integer> context = cc.getCC();

        for (Map.Entry<Pair<String, Integer>, Product> entry : map.entrySet()) {
            Pair<String, Integer> key = entry.getKey();
            Product value = entry.getValue();
            // if context contains element_id
            if (cc.getCC().containsKey(key.getKey()) && cc.getCC().get(key.getKey()).equals(key.getValue())) {
                filteredMap.put(key, value);
            }
        }
        return filteredMap.values();
    }

    /**
     * Add a new element to the map, keeping the causal context
     *
     * @param element_id Identifier of the element
     * @param element    Element to be added
     */
    public void add(String element_id, Product element) {
        Pair<String, Integer> dot = cc.makeDot(element_id);
        map.put(dot, element);
    }

    /**
     * Remove an element from the map, keeping the causal context
     *
     * @param element_id Identifier of the element
     */
    public void rm(String element_id) {
        Iterator<Map.Entry<Pair<String, Integer>, Product>> iterator = map.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry<Pair<String, Integer>, Product> entry = iterator.next();
            Pair<String, Integer> key = entry.getKey();
            if (key.getKey().equals(element_id)) {
                iterator.remove();
            }
        }
    }

    /**
     * Get an element from the map
     *
     * @param element_id Identifier of the element
     * @return Element
     */
    public Product get(String element_id) {
        for (Map.Entry<Pair<String, Integer>, Product> entry : map.entrySet()) {
            Pair<String, Integer> key = entry.getKey();
            if (key.getKey().equals(element_id)) {
                return entry.getValue();
            }
        }
        return null;
    }

    /**
     * Join two AWORMaps
     *
     * @param awormap AWORMap to join
     */
    public void join(AWORMap awormap) {
        Iterator<Map.Entry<Pair<String, Integer>, Product>> iterator = map.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry<Pair<String, Integer>, Product> entry = iterator.next();
            Pair<String, Integer> key = entry.getKey();
            if (!awormap.getMap().containsKey(key) || !awormap.getCC().dotIn(key)) {
                iterator.remove();
            }
        }

        for (Map.Entry<Pair<String, Integer>, Product> entry : awormap.getMap().entrySet()) {
            Pair<String, Integer> key = entry.getKey();
            Product value = entry.getValue();
            if (!cc.dotIn(key)) {
                map.put(key, value);
            }
        }

        cc.join(awormap.getCC());
    }

    public void fromJSON(String id, JSONObject products) {
        this.id = id;

        for (Object product : products.getJSONArray("map")) {
            JSONObject jProduct = (JSONObject) product;
            Pair<String, Integer> dot = new Pair<>(jProduct.getString("name"), jProduct.getInt("context"));
            Product newProduct = new Product(jProduct.getString("name"), id);
            newProduct.fromJSON(jProduct);

            map.put(dot, newProduct);
        }

        for (Object dot : products.getJSONObject("context").getJSONArray("cc")) {
            JSONArray jDot = (JSONArray) dot;
            cc.getCC().put(jDot.getString(0), jDot.getInt(1)); // CHECKME
        }

        if (!products.getJSONObject("context").getJSONArray("dc").isEmpty()) {
            for (Object dot : products.getJSONObject("context").getJSONArray("dc")) {
                JSONArray jDot = (JSONArray) dot;
                Pair<String, Integer> pairDot = new Pair<>(jDot.getString(0), jDot.getInt(1));
                cc.insertDot(pairDot, false);
                // TESTME
            }
        }
    }

    public JSONObject toJSON() {
        JSONObject res = new JSONObject();
        ArrayList<JSONObject> map = new ArrayList<JSONObject>();
        res.put("context", new HashMap<>());

        for (Map.Entry<Pair<String, Integer>, Product> entry : this.map.entrySet()) {
            Product product = entry.getValue();
            JSONObject productJSON = new JSONObject();
            productJSON.put("name", entry.getKey().getKey());
            productJSON.put("context", entry.getKey().getValue());
            productJSON.put("counter", product.toJSON());
            map.add(productJSON);
        }

        res.put("map", new JSONArray(map));
        res.put("context", this.cc.toJSON());

        return res;

    }
}

File: local/src/lib/crdt/awormap.js
// Optimized Add-Wins Observed-Remove Map

import { DotContext } from "../crdt/dotcontext";
import { Product } from "../product";

/**
 * Add-Wins Observed-Remove Map (AWORMap) - A map that stores [element_id, dot_value] =>
 * element and keeps causal context using a DotContext that maps element_id => dot_value.
 * This is optimized thanks to the DotContext implementation.
 */
export class AWORMap {
    #id; // this is most likely deprecated
    /**
     * Map that stores [element_id, dot_value] => element
     */
    #map = new Map();

    /**
     * DotContext that stores element_id => dot_value
     */
    #cc = new DotContext();

    constructor(id) {
        this.#id = id;
    }

    /**
     *
     * @returns {Map} map
     */
    getMap() {
        const res = this.#map;
        return res;
    }

    /**
     *
     * @returns {DotContext} causal context
     */
    getCC() {
        const res = this.#cc;
        return res;
    }

    elements() {
        return this.#map.keys();
    }

    /**
     * Returns the elements of the map that are in the causal context
     *
     * @returns {IterableIterator} elements
     */
    values() {
        const cc = this.#cc.getCC();
        return new Map(
            [...this.#map].filter(([key, value]) => cc.get(key[0]) == key[1]),
        ).values();
    }

    /**
     * Returns a map of the elements that are in the causal context
     *
     * DEPRECATED: #map is now correct
     *
     * @returns {Map} elements
     */
    flatten() {
        const cc = this.#cc.getCC();
        return new Map(
            [...this.#map].filter(([key, value]) => cc.get(key[0]) == key[1]),
        );
    }

    /**
     * Adds a new element to the map, keeping the causal context
     *
     * @param {*} element_id Identifier of the element
     * @param {*} element Element to be added
     */
    add(element_id, element) {
        const dot = this.#cc.makeDot(element_id);
        for (const [key, value] of this.#map) {
            if (key[0] == element_id) {
                this.#map.delete(key);
            }
        }
        this.#map.set([dot[0], dot[1]], element); // [product.getName(), 1] => Product
    }

    /**
     * Removes an element from the map, keeping the causal context
     *
     * @param {*} element_id Identifier of the element
     */
    rm(element_id) {
        const dot = this.#cc.makeDot(element_id);

        for (const [key, value] of this.#map) {
            if (key[0] == element_id) {
                this.#map.delete(key);
            }
        }
    }

    get(element_id) {
        for (const [key, value] of this.#map) {
            if (key[0] == element_id) return value;
        }
        return undefined;
    }

    /**
     * Checks if [key,value] is in the map
     * Note: I'm aware this is preposterous code but apparently two arrays ["banana", 1] and ["banana", 1] are not equal
     *
     * @param {Array} key
     * @returns {Boolean}
     */
    hasKey(key) {
        for (const [_key, value] of this.#map) {
            console.log(
                _key,
                key,
                value,
                _key[0] == key[0] && _key[1] == key[1],
            );
            if (_key[0] == key[0] && _key[1] == key[1]) return true;
        }
        return false;
    }

    /**
     * Join two AWORMaps
     *
     * @param {AWORMap} awormap
     */
    join(awormap) {
        this.#map = new Map(
            [...this.#map].filter(
                ([key, value]) =>
                    awormap.hasKey(key) || !awormap.getCC().dotIn(key),
            ),
        );

        for (const [key, value] of awormap.getMap()) {
            if (!this.#cc.dotIn(key)) {
                this.#map.set(key, value);
            }
        }

        this.#cc.join(awormap.getCC());
    }

    fromJSON(id, products) {
        this.#id = id;

        for (const jproduct of products.map) {
            const dot = [jproduct.name, jproduct.context];
            // TODO: Move this to the ShoppingList class
            const product = new Product();

            product.fromJSON(jproduct);

            this.#map.set(dot, product);
        }

        // FIXME: this seems to be an edge case, where due to the fact that we are reconstructing a map from a JSON, we need to
        // insert the dots in the CC, but using insertDot() won't work due to how the compacting works, there may be a better way
        // to do this, or not it could be fine as it is as it doesn't seem to be a common issue
        for (const dot of products.context.cc)
            this.#cc.getCC().set(dot[0], dot[1]);

        if (products.context.dc.length > 0)
            for (const dot of products.context.dc)
                this.#cc.insertDot(dot, false);
    }

    toJSON() {
        const res = {
            map: [],
            context: {},
        };

        for (const [key, value] of this.#map) {
            const product = value.toJSON();
            product.name = key[0];
            product.context = key[1];
            res.map.push(product);
        }

        res.context = this.#cc.toJSON();

        return res;
    }

    toFrontendJSON() {
        const res = [];

        for (const [key, value] of this.#map) {
            res.push({
                name: value.getName(),
                count: value.value(),
            });
        }

        return res;
    }
}
